### PR TITLE
Document count, sort features and document search query optimization

### PR DESCRIPTION
**Pull request by niio972**
Add document count, sort features . Also  query optimization has been done for document service.
Need these features to work on References #112, #87 .

Status : 

- [X]  developped
- [X]  tested
- [ ]  integrated

### CODE REVIEW

**Comment by annetireau** [comment]
Rename class and some coding style adjustment are needed.
I did not find any bug in the web service

File: phis2-ws/src/main/java/phis2ws/service/resources/DocumentResourceService.java [hunk 1]
```diff
@@ -356,7 +362,8 @@ public Response getDocumentsMetadataBySearch(
         @ApiParam(value = "Search by creation date", example = DocumentationAnnotation.EXAMPLE_DOCUMENT_CREATION_DATE) @QueryParam("creationDate") @Date(DateFormat.YMD) String creationDate,
         @ApiParam(value = "Search by extension", example = DocumentationAnnotation.EXAMPLE_DOCUMENT_EXTENSION) @QueryParam("extension") String extension,
         @ApiParam(value = "Search by concerned item", example = DocumentationAnnotation.EXAMPLE_EXPERIMENT_URI) @QueryParam("concernedItem") @URL String concernedItem,
-        @ApiParam(value = "Search by status", example = DocumentationAnnotation.EXAMPLE_DOCUMENT_STATUS) @QueryParam("status") String status) {
+        @ApiParam(value = "Search by status", example = DocumentationAnnotation.EXAMPLE_DOCUMENT_STATUS) @QueryParam("status") String status,
+        @ApiParam(value = "Order the result by date", allowableValues = DocumentationAnnotation.EXAMPLE_ORDER_ALLOWABLE_VALUES) @QueryParam("order") @Order String order) {
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/resources/DocumentResourceService.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/resources/DocumentResourceService.java [hunk 2]
```diff
@@ -392,6 +399,10 @@ public Response getDocumentsMetadataBySearch(
         if (status != null) {
             documentDao.status = status;
         }
+        // Order the result list by date (DESC by default)
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/resources/DocumentResourceService.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/resources/validation/interfaces/Order.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+//******************************************************************************
+//                                       Order.java
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/resources/validation/interfaces/Order.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/resources/validation/interfaces/Order.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * Interface to check that the given sort parameter is one of thoses expected
+ * @see OrderValidator
+ * @author Arnaud Charleroy<arnaud.charleroy@inra.fr>, Morgane Vidal <morgane.vidal@inra.fr>
+ */
+@Target(value={METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER})
+@Retention(RUNTIME)
+@Constraint(validatedBy = OrderValidator.class)
+@Documented
+public @interface Order {
+    String message() default "is not a right sort parameter";
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/resources/validation/interfaces/Order.java]: Modified

File: phis2-ws/src/main/java/phis2ws/service/utils/sparql/SPARQLStringBuilder.java [hunk 5]
```diff
@@ -1,24 +1,21 @@
...
 /**
- * Classe abstraite qui permet la création de constructeur qui facilite
- * l'écriture de requête SPARQL à partir de chaîne de caractères
- *
- * @author Arnaud Charleroy
+ * Abstract class which provides common methods and attributes 
+ * to manipulate SPARQL query and update.
+ * @update [No information] October, 2016 : no explanation
+ * @update [Arnaud Charleroy] 10 September, 2018 : Add "And" and "Or" filter choices
+ * @author Arnaud Charleroy <arnaud.charleroy@inra.fr>
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/utils/sparql/SPARQLStringBuilder.java]: Okay I will put author before

File: phis2-ws/src/main/java/phis2ws/service/resources/DocumentResourceService.java [hunk 6]
```diff
@@ -84,6 +81,13 @@
 import phis2ws.service.view.brapi.form.ResponseFormPOST;
 import phis2ws.service.view.model.phis.Document;
 
+/**
+ * Represents the documents service.
+ * @author Arnaud Charleroy <arnaud.charleroy@inra.fr>, Morgane Vidal <morgane.vidal@inra.fr>
+ * @update [Morgane Vidal] March, 2017 : no explanation
+ * @update [Arnaud Charleroy] 07, September 2018 : add sort feature, query optimization (limit , offset, group_concat)
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/resources/DocumentResourceService.java]: Can you be more precise ?

File: phis2-ws/src/main/java/phis2ws/service/configuration/Orders.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+//******************************************************************************
+//                                       Orders.java
+// SILEX-PHIS
+// Copyright © INRA 2018
+// Creation date: 07, Sep 2018
+// Contact: arnaud.charleroy@inra.fr, morgane.vidal@inra.fr, anne.tireau@inra.fr, 
+// pascal.neveu@inra.fr
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/configuration/Orders.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/dao/sesame/DocumentDaoSesame.java [hunk 8]
```diff
@@ -54,23 +40,66 @@
...
+    /**
+     * Sort document by date
+     * Allowable values : asc, desc
+     */
+    public String order;
+    
     //List of the elements concerned by the document
     public List<String> concernedItemsUris = new ArrayList<>();
+    public static final String CONCERN = "concern";
+    public static final String CONCERN_TYPE = "typeConcern";
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/dao/sesame/DocumentDaoSesame.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/dao/sesame/DocumentDaoSesame.java [hunk 9]
```diff
@@ -372,33 +431,38 @@ protected SPARQLQueryBuilder prepareSearchQuery() {
...
-        sparqlQuery.appendTriplet(uriDocument, TRIPLESTORE_RELATION_COMMENT, "?comment", null);
+        if (order != null) {
+            sparqlQuery.appendOrderBy(order.toUpperCase() + "(?" + CREATION_DATE + ")");
+        }else{
+            // Use by default DESC if the order parameter is null
+            sparqlQuery.appendOrderBy(Orders.DESC.toString().toUpperCase() + "(?" + CREATION_DATE + ")");
+        }
+        
+        sparqlQuery.appendLimit(this.getPageSize());
+        sparqlQuery.appendOffset(this.getPage()* this.getPageSize());
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/dao/sesame/DocumentDaoSesame.java]: okay

File: phis2-ws/src/main/java/phis2ws/service/dao/sesame/DocumentDaoSesame.java [hunk 10]
```diff
@@ -407,25 +471,73 @@ private SPARQLQueryBuilder prepareSearchConcernQuery(String uriDocument) {
...
+     * @return query generated with the searched parameters
+     */
+    private SPARQLQueryBuilder prepareCount() {
+        SPARQLQueryBuilder query = this.prepareSearchQuery();
+        query.clearSelect();
+        query.clearLimit();
+        query.clearOffset();
+        query.clearGroupBy();
+        query.clearOrderBy();
+        query.appendSelect("(count(distinct ?" + URI + ") as ?" + COUNT_ELEMENT_QUERY + ")");
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/dao/sesame/DocumentDaoSesame.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/view/brapi/form/ResponseFormDocumentMetadata.java [hunk 11]
```diff
@@ -33,4 +39,21 @@ public ResponseFormDocumentMetadata(int pageSize, int currentPage, ArrayList<Doc
...
+    /**
+     * Initiate metadata and result response fields
+     * @param pageSize results per page
+     * @param currentPage current page
+     * @param list results list
+     * @param paginate need to be paginate or not
+     * @param totalCount the number of the items returned by the query
+     */
+    public ResponseFormDocumentMetadata(int pageSize, int currentPage, ArrayList<Document> list, boolean paginate, int totalCount) {
+        metadata = new Metadata(pageSize, currentPage, totalCount);
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/view/brapi/form/ResponseFormDocumentMetadata.java]: Can you be more precise, please ?

File: phis2-ws/src/main/java/phis2ws/service/utils/sparql/SPARQLStringBuilder.java [hunk 12]
```diff
@@ -115,10 +112,41 @@ public void stringifyObjectForm() {
...
+    
+    /**
+     * @example
+     * FILTER ( (regex(STR(?creator), 'admin', 'i')) && (regex(STR(?title), 'liste', 'i')) ) 
+     * @param filter the filter which will be applied e.g. regex(STR(?creator), 'admin', 'i')
+     */
+    public void appendAndFilter(String filter) {
+        if(this.filter != ""){
+             this.filter += " && (" + filter  + ")";
+        }else{
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/utils/sparql/SPARQLStringBuilder.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/dao/sesame/DocumentDaoSesame.java [hunk 13]
```diff
@@ -372,33 +431,38 @@ protected SPARQLQueryBuilder prepareSearchQuery() {
...
-     */
-    private SPARQLQueryBuilder prepareSearchCommentQuery(String uriDocument) {
-        SPARQLQueryBuilder sparqlQuery = new SPARQLQueryBuilder();
-        sparqlQuery.appendDistinct(true);
-        sparqlQuery.appendGraph(TRIPLESTORE_GRAPH_DOCUMENT);
-        sparqlQuery.appendSelect("?comment");
-        sparqlQuery.appendTriplet(uriDocument, TRIPLESTORE_RELATION_COMMENT, "?comment", null);
+        if (order != null) {
+            sparqlQuery.appendOrderBy(order.toUpperCase() + "(?" + CREATION_DATE + ")");
+        }else{
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/dao/sesame/DocumentDaoSesame.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/utils/sparql/SPARQLStringBuilder.java [hunk 14]
```diff
@@ -115,10 +112,41 @@ public void stringifyObjectForm() {
...
+        this.filter +=  "(" + filter + ")";
+    }
+    
+    /**
+     * @example
+     * FILTER ( (regex(STR(?creator), 'admin', 'i')) && (regex(STR(?title), 'liste', 'i')) ) 
+     * @param filter the filter which will be applied e.g. regex(STR(?creator), 'admin', 'i')
+     */
+    public void appendAndFilter(String filter) {
+        if(this.filter != ""){
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/utils/sparql/SPARQLStringBuilder.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/utils/sparql/SPARQLStringBuilder.java [hunk 15]
```diff
@@ -115,10 +112,41 @@ public void stringifyObjectForm() {
...
+        }
     }
 
+    /**
+     * @example
+     * FILTER ( (regex(STR(?creator), 'admin', 'i')) || (regex(STR(?title), 'liste', 'i')) ) 
+     * @param filter the filter which will be applied e.g. regex(STR(?creator), 'admin', 'i')
+     */
+    public void appendOrFilter(String filter) {
+        if(this.filter != ""){
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/utils/sparql/SPARQLStringBuilder.java]: Done

File: phis2-ws/src/main/java/phis2ws/service/configuration/Orders.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
+//******************************************************************************
+//                                       Orders.java
```
[the remark this answers is not in the document]
Comment (niio972) [reply phis2-ws/src/main/java/phis2ws/service/configuration/Orders.java]: Done

**Pull request merged**